File: BOJ/1463_2_1.py
'''
    1로 만들기 (BottomUp)
    https://www.acmicpc.net/problem/1463

    파이썬은 스택 메모리 제한이 있어서 재귀를 많이 할 것 같다면
    BottomUp 방식으로 짜야한다.

'''
from sys import stdin
MAX = 1000000 + 1
dp = [MAX] * MAX

input = stdin.readline
N = int(input())
dp[1] = 0  # 기저
for i in range(N):
    dp[i+1] = min(dp[i+1], dp[i] + 1)
    if i * 2 < MAX:
        dp[i*2] = min(dp[i*2], dp[i] + 1)
    if i * 3 < MAX:
        dp[i*3] = min(dp[i*3], dp[i] + 1)

print(dp[N])

File: BOJ/1463_2_1.cpp
// 1로 만들기 (TopDown)
// https://www.acmicpc.net/problem/1463
#include <iostream>
#include <algorithm>
using namespace std;
int N;
const int MAX = 1000000 + 1;
int dp[MAX];

int TopDown(int n)
{
    // 기저
    if (n == 1)
        return 0;
    // 이미 구한 경우
    if (dp[n] != -1)
        return dp[n];

    int result = TopDown(n - 1) + 1;
    if (n % 3 == 0)
        result = min(result, TopDown(n / 3) + 1);
    else if (n % 2 == 0)
        result = min(result, TopDown(n / 2) + 1);
    dp[n] = result;
    return result;
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    cin >> N;
    fill_n(dp, MAX, -1);
    cout << TopDown(N) << '\n';
}